doc/warm-reboot/NSF_Manager_HLD.md

```diff
@@ -0,0 +1,280 @@
+
+
+#### Shutdown Orchestration
+
+During warm reboot, NSF Manager will orchestrate switch shutdown in a multi-phased approach that ensures that the switch stack is in a stable state before the intent/state is saved. It will perform shutdown in the following phases:
+
+
+
+*   Phase 1: Freeze components & wait for switch quiescence
+*   Phase 2: Perform state verification (optional)
+*   Phase 3: Trigger checkpointing
+*   Phase 4: Prepare and perform reboot
+
+
+##### Phase 1: Freeze Components & Wait for Switch Quiescence
+
+
+
+![alt_text](img/freeze.png)
+
+NSF Manager will send freeze notification to all registered components and will wait for the quiescence of only those components that have set _freeze = true_. Upon receiving the freeze notification, the component will complete processing the current request queue and stop generating new intents. Stopping new intents from being generated means that boundary components should stop processing requests from external components (external events) and all components should stop their periodic timers that generate new requests (internal events). For example:
+
+
+
+*   UMF will stop listening to new gRPC requests from the controller.
+*   P4RT will stop listening to new gRPC requests from the controller and stop packet I/O.
+*   xcvrd will stop listening to transceiver updates such as module presence.
+*   Syncd will stop listening to link state events from the BCM chip.
+*   Orchagent will stop periodic internal timers.
+*   BGP will stop exchanging packets with the peers.
+*   Teamd will stop exchanging LACP PDUs with the peers.
+
+After all components have been freezed, the switch would eventually reach a state wherein each component stops generating new events and thus the switch becomes quiescent. This is because:
+
+
+
+*   Switch boundaries that generate new events have been stopped.
+*   All timers that generate new events have been stopped.
+*   All components have completed processing their pending requests and thus there are no in-flight messages.
+
+After receiving the freeze notification, the components will update their quiescent state in STATE DB when they receive a new request (i.e. they are no longer quiescent) and when they complete processing their current request queue (i.e. they become quiescent). NSF Manager will monitor the quiescent state of all components in STATE DB to determine that the switch has become quiescent and thus further state changes won’t occur in the switch. If all components are in quiescent state then NSF Manager will declare that the switch has become quiescent and thus the switch has attained its final state. NSF Manager will wait for a period of time for the switch to become quiescent after which it will determine that warm reboot failed and abort the warm reboot operation.
```

Comment:
> NSF Manager will wait for a period of time for the switch to become quiescent after which it will determine that warm reboot failed and abort the warm reboot operation.

The design should also account for restoration of components state (unfreeze) in the event of failure.

Reply:
Does SONiC currently support unfreeze of applications in case of failures?